src/main/java/jcifs/smb1/smb1/SmbSession.java

```diff
@@ -0,0 +1,476 @@
+                 * assert that the tree connect path uses the netbios hostname.
+                 */
+                tcax.path = "\\\\" + netbiosName + "\\IPC$";
+            }
+        }
+
+        request.uid = uid;
+        request.auth = auth;
+        try {
+            transport.send( request, response );
+        } catch (SmbException se) {
+            if (request instanceof SmbComTreeConnectAndX) {
+                logoff(true);
+            }
+            request.digest = null;
+            throw se;
+        }
+}
+    }
+    void sessionSetup( ServerMessageBlock andx,
+                ServerMessageBlock andxResponse ) throws SmbException {
+synchronized (transport()) {
+        NtlmContext nctx = null;
+        SmbException ex = null;
+        SmbComSessionSetupAndX request;
+        SmbComSessionSetupAndXResponse response;
+        byte[] token = new byte[0];
+        int state = 10;
+
+        while (connectionState != 0) {
+            if (connectionState == 2 || connectionState == 3) // connected or disconnecting
+                return;
+            try {
+                transport.wait();
+            } catch (InterruptedException ie) {
+                throw new SmbException(ie.getMessage(), ie);
+            }
+        }
+        connectionState = 1; // trying ...
+
+        try {
+            transport.connect();
+
+            /*
+             * Session Setup And X Request / Response
+             */
+    
+            if( transport.log.level >= 4 )
+                transport.log.println( "sessionSetup: accountName=" + auth.username + ",primaryDomain=" + auth.domain );
+    
+            /* We explicitly set uid to 0 here to prevent a new
+             * SMB_COM_SESSION_SETUP_ANDX from having it's uid set to an
+             * old value when the session is re-established. Otherwise a
+             * "The parameter is incorrect" error can occur.
+             */
+            uid = 0;
+    
+            do {
+                switch (state) {
+                    case 10: /* NTLM */
+                        if (auth != NtlmPasswordAuthentication.ANONYMOUS &&
+                                transport.hasCapability(SmbConstants.CAP_EXTENDED_SECURITY)) {
+                            state = 20; /* NTLMSSP */
+                            break;
+                        }
+    
+                        request = new SmbComSessionSetupAndX( this, andx, auth );
+                        response = new SmbComSessionSetupAndXResponse( andxResponse );
+    
+                        /* Create SMB signature digest if necessary
+                         * Only the first SMB_COM_SESSION_SETUP_ANX with non-null or
+                         * blank password initializes signing.
+                         */
+                        if (transport.isSignatureSetupRequired( auth )) {
+                            if( auth.hashesExternal && NtlmPasswordAuthentication.DEFAULT_PASSWORD != NtlmPasswordAuthentication.BLANK ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.